components/script/dom/blob.rs

```
@@ -118,35 +152,70 @@ impl Blob {
-        Ok(Blob::new(global, slice, &blobPropertyBag.get_typestring()))
+        let slice = DataSlice::from_bytes(bytes);
+        Ok(Blob::new(global, BlobImpl::new_from_slice(slice), &blobPropertyBag.get_typestring()))
     }
 
-    pub fn get_data(&self) -> &DataSlice {
-        &self.data
+    /// Get a slice to inner data, this might incur synchronous read and caching
+    /// XXX: it is still possible to bring down the whole program, but the API might be
+    ///      intuitive to use (since we read FS and cache in a transparent style)
+    pub fn get_slice(&self) -> DataSlice {
+        match self.blob_impl {
+            BlobImpl::File(ref id, ref slice) => {
+                match *slice.borrow() {
+                    Some(ref s) => s.clone(),
+                    None => {
+                        let global = self.global();
+                        let s = read_file(global.r(), id.clone()).expect("Read file in get_slice failed");
+                        *slice.borrow_mut() = Some(s.clone()); // Cached
+                        s
+                    }
+                }
+            }
+            BlobImpl::Memory(ref s) => s.clone()
+        }
     }
 }
 
-pub fn blob_parts_to_bytes(blobparts: Vec<BlobOrString>) -> Vec<u8> {
-    blobparts.iter().flat_map(|blobpart| {
-            match blobpart {
-                &BlobOrString::String(ref s) => {
-                    UTF_8.encode(s, EncoderTrap::Replace).unwrap()
-                },
-                &BlobOrString::Blob(ref b) => {
-                    b.get_data().get_bytes().to_vec()
-                },
-            }
-        }).collect::<Vec<u8>>()
+fn read_file(global: GlobalRef, id: Uuid) -> Result<DataSlice, ()> {
+    let file_manager = global.filemanager_thread();
+    let (chan, recv) = ipc::channel().map_err(|_|())?;
+    let _ = file_manager.send(FileManagerThreadMsg::ReadFile(chan, id));
+    let result = recv.recv().unwrap();
```

This should propagate an error